tasmota/tasmota_xsns_sensor/xsns_110_max17043.ino

```diff
@@ -0,0 +1,130 @@
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef USE_I2C
+#ifdef USE_MAX17043
+
+#define XI2C_83       83    // See I2CDEVICES.md
+
+#define SENSOR_NAME    "MAX17043"
+
+#include "DFRobot_MAX17043.h"
+
+/*********************************************************************************************\
+ * MAX17043 fuel-gauge for 3.7 Volt Lipo batteries
+ *
+ * Battery voltage in Volt and State Of Charge (SOC) in percent are published via MQTT
+ * 
+ * The alert flag and alert threshold are not required for MQTT, the alert pin is not used 
+ * by this sensor driver.
+ * 
+ * Wirering and other information:
+ * 
+ * \lib\lib_i2c\DFRobot_MAX17043\resources
+ * 
+ * Tested module(s):
+ * 
+ * https://www.dfrobot.com/product-1734.html
+ * 
+ * Not yet tested module(s):
+ * 
+ * https://www.aliexpress.us/item/2251832479401925.html 
+ * 
+ \*********************************************************************************************/
+
+#define XSNS_109        109
+
+const char *mqttId = "MAX17043";
+
+DFRobot_MAX17043     gauge; // Class to read from MAX17043
+
+struct MAX17043 
+{
+  float voltage = 0.0;      // Battery voltage in Volt
+  float percentage = 0.0;   // Battery remaining capacity in percent
+} *max17043 = nullptr;
+
+ /*********************************************************************************************/
+
+void Max17043Init(void) {
+
+  if (I2cSetDevice(MAX17043_ADDRESS)) { 
+    I2cSetActiveFound(MAX17043_ADDRESS, "MAX17043");
+    if (gauge.begin() == 0) {
+      max17043 = (MAX17043 *)calloc(1, sizeof(struct MAX17043));
+    }
+  }
+}
+
+void Max17043Read(void) {
+
+  float percentage = 0.0;
+  
+  max17043->voltage = gauge.readVoltage()/1000.0;
+
+  // During charging the percentage might be (slightly) above 100%. To avoid stange numbers
+  // in the statistics we the percentage provided by this driver will not go above 100%
+  percentage = gauge.readPercentage();
+  if (percentage > 100.0) {
+    max17043->percentage = 100.0;
+  }
+  else {
+    max17043->percentage = percentage;
+  }
+}
+
+void Max17043Json(void) {
+  ResponseAppend_P(PSTR(",\"%s\":{\"" D_JSON_VOLTAGE "\":%3_f,\"" D_JSON_BATTPERCENT "\":%2_f}"), mqttId, &max17043->voltage, &max17043->percentage );
+}
+
+#ifdef USE_WEBSERVER
+void Max17043Show(void) {
+  WSContentSend_PD(PSTR("{s}%s " D_VOLTAGE "{m}%1_f" D_UNIT_VOLT "{e}"), SENSOR_NAME, &max17043->voltage);
+  WSContentSend_PD(PSTR("{s}%s " D_BATTERY_CHARGE "{m}%1_f %% {e}"), SENSOR_NAME, &max17043->percentage);
+}  
+#endif  // USE_WEBSERVER
+
+/*********************************************************************************************\
+ *  Interface
+\*********************************************************************************************/
+
+bool Xsns109(uint32_t function) {
+if (!I2cEnabled(MAX17043_ADDRESS)) { return false; } 
+
+  switch (function) {
+    case FUNC_INIT:
+      Max17043Init(); 
+      break;
+    case FUNC_EVERY_SECOND:
```

Comment:
This construct needs changing! As it stands even with no sensr connected it keeps executing the following cases. You'll need to change the FUNC_INIT case to something disabling the below cases if no sensor found. Pls look at other sensor drivers for examples.

Reply:
A typical method is to have a global pointer to all needed driver data that is initialized as nullptr
If you detect the device at INIT then you can 'calloc()' for your data and assign to that pointer
This provides 2 things
- do not use any fixed allocated RAM if the device is not detected
- a simple value to test against for all calls but INIT

Reply:
Should be solved now by [2077d08](https://github.com/arendst/Tasmota/pull/18788/commits/2077d08ef75ec3e2138bb5a413e4d11ba0a6e111)